x/wasm/internal/keeper/staking_test.go

```diff
@@ -404,6 +406,106 @@ func TestReinvest(t *testing.T) {
+
+	// upload mask code
+	maskCode, err := ioutil.ReadFile("./testdata/reflect.wasm")
+	require.NoError(t, err)
+	maskID, err := keeper.Create(ctx, creator, maskCode, "", "", nil)
+	require.NoError(t, err)
+	require.Equal(t, uint64(2), maskID)
+
+	// creator instantiates a contract and gives it tokens
+	maskAddr, err := keeper.Instantiate(ctx, maskID, creator, nil, []byte("{}"), "mask contract 2", nil)
+	require.NoError(t, err)
+	require.NotEmpty(t, maskAddr)
+
+	// STEP 3: now, let's reflect some queries.
+	// now, let's reflect a smart query into the x/wasm handlers and see if we get the same result
+	reflectValidatorsQuery := MaskQueryMsg{Chain: &ChainQuery{Request: &wasmTypes.QueryRequest{Staking: &wasmTypes.StakingQuery{
+		Validators: &wasmTypes.ValidatorsQuery{},
+	}}}}
+	reflectValidatorsBin := buildMaskQuery(t, &reflectValidatorsQuery)
+	res, err := keeper.QuerySmart(ctx, maskAddr, reflectValidatorsBin)
+	require.NoError(t, err)
+	// first we pull out the data from chain response, before parsing the original response
+	var reflectRes ChainResponse
+	mustParse(t, res, &reflectRes)
+	var validatorRes wasmTypes.ValidatorsResponse
+	mustParse(t, reflectRes.Data, &validatorRes)
+	require.Equal(t, 1, len(validatorRes.Validators))
+	valInfo := validatorRes.Validators[0]
+	// Note: this ValAddress not AccAddress, may change with #264
+	require.Equal(t, valAddr.String(), valInfo.Address)
+	require.Contains(t, valInfo.Commission, "0.100")
+	require.Contains(t, valInfo.MaxCommission, "0.200")
+	require.Contains(t, valInfo.MaxChangeRate, "0.010")
+
+	// test to get all my delegations
+	reflectAllDelegationsQuery := MaskQueryMsg{Chain: &ChainQuery{Request: &wasmTypes.QueryRequest{Staking: &wasmTypes.StakingQuery{
+		AllDelegations: &wasmTypes.AllDelegationsQuery{
+			Delegator: contractAddr.String(),
+		},
+	}}}}
+	reflectAllDelegationsBin := buildMaskQuery(t, &reflectAllDelegationsQuery)
+	res, err = keeper.QuerySmart(ctx, maskAddr, reflectAllDelegationsBin)
+	require.NoError(t, err)
+	// first we pull out the data from chain response, before parsing the original response
+	mustParse(t, res, &reflectRes)
+	var allDelegationsRes wasmTypes.AllDelegationsResponse
+	mustParse(t, reflectRes.Data, &allDelegationsRes)
+	require.Equal(t, 1, len(allDelegationsRes.Delegations))
+	delInfo := allDelegationsRes.Delegations[0]
+	fmt.Printf("%#v\n", delInfo)
+	// Note: this ValAddress not AccAddress, may change with #264
+	require.Equal(t, valAddr.String(), delInfo.Validator)
+	// note this is not bob (who staked to the contract), but the contract itself
+	require.Equal(t, contractAddr.String(), delInfo.Delegator)
+	// this is a different Coin type, with String not BigInt, compare field by field
+	require.Equal(t, funds[0].Denom, delInfo.Amount.Denom)
+	require.Equal(t, funds[0].Amount.String(), delInfo.Amount.Amount)
+
+	// TODO: more delegation queries
+	// BondedDenom -> Atom
+	// Delegation -> With detailed info
+}
+
```

Comment:
pp: there was quite some effort to setup the test. I was wondering if a staking keeper mock would be an easier fit or what was your intention?

Reply:
We need to see how the query callbacks from the contracts really work.

I have a contract that will take the user input and use that to trigger a query via `Querier`. This allows us to see what a contract would see. And we need to see what would happen against a real staking keeper (as there are some bugs... we want to fix how those work, full stack for compatibility).